### PR TITLE
[LoongArch64]: Add loongarch64 architecture to roslyn

### PR DESCRIPTION
**Pull request by chenguohui**
To support target runtime "linux-loongarch64" in .NET SDK.

The build command looks like:

`dotnet build -c Release -r linux-loongarch64`

### CODE REVIEW

**Comment by chenguohui** [comment]
@shushanhf

**Comment by shushanhf** [comment]
qiao@Lap5K:~/Downloads/hello8$    dotnet publish -p:PublishReadyToRun=true  hello8.csproj 
适用于 .NET MSBuild 版本 17.8.5+b5265ef37
  Determining projects to restore...
  Restored /home/qiao/Downloads/hello8/hello8.csproj (in 814 ms).
  hello8 -> /home/qiao/Downloads/hello8/bin/Release/net8.0/linux-loongarch64/hello8.dll
/usr/share/dotnet/sdk/8.0.107/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(396,5): error NETSDK1095: 所选目标平台或体系结构不支持优化程序集的性能。请验证确保你在使用受支持的运行时标识符，或者将 PublishReadyToRun 属性设置为 false。 [/home/qiao/Downloads/hello8/hello8.csproj]

**Comment by jaredpar** [comment]
Is there an issue, link, etc ... that explains the motivation for this change?

**Comment by jjonescz** [comment]
https://github.com/dotnet/roslyn/issues/76065

**Comment by jkotas** [comment]
> To support target runtime "linux-loongarch64" in .NET SDK.

Could you please elaborate on why this change is needed to support loongarch64 in .NET SDK?

**Comment by chenguohui** [comment]
> Could you please elaborate on why this change is needed to support loongarch64 in .NET SDK?

I want to improve the performance on LoongArch64 platform. And this implementation can help improving the performance by generate optimized DLLs.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 1]
```diff
@@ -3119,7 +3119,7 @@ A catch() block after a catch (System.Exception e) block can catch non-CLS excep
     <value>A namespace declaration cannot have modifiers or attributes</value>
   </data>
   <data name="ERR_BadPlatformType" xml:space="preserve">
-    <value>Invalid option '{0}' for /platform; must be anycpu, x86, Itanium, arm, arm64 or x64</value>
+    <value>Invalid option '{0}' for /platform; must be anycpu, x86, Itanium, arm, arm64, loongarch64 or x64</value>
```
Comment (jkotas) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:3122]: This and other changes related to loongarch64 value of platform switch should be reverted too.

File: src/Compilers/Core/Portable/DiaSymReader/SymUnmanagedFactory.cs [hunk 2]
```diff
@@ -20,6 +20,7 @@ internal static class SymUnmanagedFactory
         private const string DiaSymReaderModuleName32 = "Microsoft.DiaSymReader.Native.x86.dll";
         private const string DiaSymReaderModuleNameAmd64 = "Microsoft.DiaSymReader.Native.amd64.dll";
         private const string DiaSymReaderModuleNameArm64 = "Microsoft.DiaSymReader.Native.arm64.dll";
+        private const string DiaSymReaderModuleNameLoongArch64 = "Microsoft.DiaSymReader.Native.loongarch64.dll";
```
Comment (jkotas) [inline src/Compilers/Core/Portable/DiaSymReader/SymUnmanagedFactory.cs:23]: This is Windows specific code. This should be reverted.

**Comment by jkotas** [comment]
> I want to improve the performance on LoongArch64 platform. And I think this implementation may help to improve the performance by generating optimized DLLs.

The Roslyn platform switch does not optimize performance in any way.

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 1 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:3122]: The issue @jjonescz linked to describes the problem:
```
chenguohui@debian-230:~/target-runtime/hello$ ~/target-runtime/dotnet-sdk-9.0.100-linux-loongarch64/dotnet bin/Release/net9.0/linux-loongarch64/hello.dll
Unhandled exception. System.BadImageFormatException: Could not load file or assembly '/home/chenguohui/target-runtime/hello/bin/Release/net9.0/linux-loongarch64/hello.dll'. An attempt was made to load a program with an incorrect format.
```

@chenguohui What specifically is wrong with the binary? It's not obvious from the information provided that there is a problem with the compiler

**Comment by jaredpar** [comment]
> I want to improve the performance on LoongArch64 platform.

As @jkotas mentioned, this PR will not have that impact. Given that are there other reasons to have this PR that we are missing?

**Comment by chenguohui** [comment]
> > I want to improve the performance on LoongArch64 platform. And I think this implementation may help to improve the performance by generating optimized DLLs.
> 
> The Roslyn platform switch does not optimize performance in any way.

All right. I really don't know the Roslyn can directly improve the performance or not.
I followed the log of generating a PE32+ dll compared `linux-loongarch64` with `linux-x64`.
On the linux-x64 platform, I found the `csc.dll` parameter is `/platform:x64` and on `linux-loongarch64` is `/platform:AnyCPU`.
At least, applied this PR, the `csc.dll` parameter changed to `/platform:loongarch64` as below.

```shell
        /home/chenguohui/target-runtime/dotnet-sdk-9.0.100-linux-loongarch64/dotnet exec "/home/chenguohui/target-runtime/dotnet-sdk-9.0.100-linux-loongarch64/sdk/9.0.100/Roslyn/bincore/csc.dll" /noconfig /unsafe- /checked- /nowarn:1701,1702,1701,1702 /fullpaths /nostdlib+ /platform:loongarch64 ... # this line is very long
```
Previously, I thought this is a switch that can open the later optimization.

**Comment by chenguohui** [comment]
> qiao@Lap5K:~/Downloads/hello8$ dotnet publish -p:PublishReadyToRun=true hello8.csproj 适用于 .NET MSBuild 版本 17.8.5+b5265ef37 Determining projects to restore... Restored /home/qiao/Downloads/hello8/hello8.csproj (in 814 ms). hello8 -> /home/qiao/Downloads/hello8/bin/Release/net8.0/linux-loongarch64/hello8.dll /usr/share/dotnet/sdk/8.0.107/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(396,5): error NETSDK1095: 所选目标平台或体系结构不支持优化程序集的性能。请验证确保你在使用受支持的运行时标识符，或者将 PublishReadyToRun 属性设置为 false。 [/home/qiao/Downloads/hello8/hello8.csproj]
> 
> The `PublishReadyToRun=true` does not support the LoongArch64 but the crossgen2.exe can generate the PE32+ formate file.

**The `error NETSDK1095` in English**

error NETSDK1095: Optimizing assemblies for performance is not supported for the selected target platform or architecture. Please verify you are using a supported runtime identifier, or set the PublishReadyToRun property to false. [/home/chenguohui/hello8/hello8.csproj]

**Comment by jkotas** [comment]
> error NETSDK1095: Optimizing assemblies for performance is not supported for the selected target platform or architectur

This issue needs should be fixed in dotnet/sdk repo. I see that you have added R2R loongarch64 support to SDK repo earlier:

https://github.com/dotnet/sdk/blob/d1e78215aad30c556e002738656f7acdd8841a3c/src/Tasks/Microsoft.NET.Build.Tasks/ResolveReadyToRunCompilers.cs#L273-L275

Was the earlier change incomplete?

**Comment by chenguohui** [comment]
> > error NETSDK1095: Optimizing assemblies for performance is not supported for the selected target platform or architectur
> 
> This issue needs should be fixed in dotnet/sdk repo. I see that you have added R2R loongarch64 support to SDK repo earlier:
> 
> https://github.com/dotnet/sdk/blob/d1e78215aad30c556e002738656f7acdd8841a3c/src/Tasks/Microsoft.NET.Build.Tasks/ResolveReadyToRunCompilers.cs#L273-L275
> 
> Was the earlier change incomplete?

Thank you for your remind!

.NET 9.0.100 doesn't exists this issue. I compared dotnet/sdk branches release/8.0.1xx and release/9.0.1xx, the release/8.0.1xx is incomplete for loongarch64.

I will try to built a new .NET 8.0.111 to fix it.

**Comment by chenguohui** [comment]
> > > I want to improve the performance on LoongArch64 platform. And I think this implementation may help to improve the performance by generating optimized DLLs.
> > 
> > 
> > The Roslyn platform switch does not optimize performance in any way.
> 
> All right. I really don't know the Roslyn can directly improve the performance or not. I followed the log of generating a PE32+ dll compared `linux-loongarch64` with `linux-x64`. On the linux-x64 platform, I found the `csc.dll` parameter is `/platform:x64` and on `linux-loongarch64` is `/platform:AnyCPU`. Applied this PR, the `csc.dll` parameter changed to `/platform:loongarch64` as below.
> 
> ```shell
>         /home/chenguohui/target-runtime/dotnet-sdk-9.0.100-linux-loongarch64/dotnet exec "/home/chenguohui/target-runtime/dotnet-sdk-9.0.100-linux-loongarch64/sdk/9.0.100/Roslyn/bincore/csc.dll" /noconfig /unsafe- /checked- /nowarn:1701,1702,1701,1702 /fullpaths /nostdlib+ /platform:loongarch64 ... # this line is very long
> ```
> 
> Previously, I thought this is a switch that can open the later optimization in the .NET SDK workflow.

Is the `csc.dll` parameter `/platform:loongarch64` useless? I'm not sure for the function.

**Comment by jcouv** [comment]
> Given that are there other reasons to have this PR that we are missing?

@chenguohui Unless a clear purpose for a roslyn change can be identified, please close the PR. Thanks

**Comment by chenguohui** [comment]
OK, thank you for your help! @jkotas @jcouv @jaredpar

**Pull request closed without merging**